ydb/core/persqueue/partition.cpp

```diff
@@ -1081,19 +988,91 @@ void TPartition::Handle(TEvPQ::TEvGetWriteInfoRequest::TPtr& ev, const TActorCon
     response->BlobsFromHead = std::move(GetReadRequestFromHead(0, 0, std::numeric_limits<ui32>::max(),
                                                                std::numeric_limits<ui32>::max(), 0, &rcount, &rsize,
                                                                &insideHeadOffset, 0));
-
-
     response->BytesWrittenGrpc = BytesWrittenGrpc.Value();
     response->BytesWrittenUncompressed = BytesWrittenUncompressed.Value();
     response->BytesWrittenTotal = BytesWrittenTotal.Value();
-
     response->MessagesWrittenTotal = MsgsWrittenTotal.Value();
     response->MessagesWrittenGrpc = MsgsWrittenGrpc.Value();
     response->MessagesSizes = std::move(MessageSize.GetValues());
 
     ctx.Send(ev->Sender, response);
 }
 
+void TPartition::WriteInfoResponseHandler(const TActorId& sender, TAutoPtr<TEvPQ::TEvGetWriteInfoResponse>&& ev, const TActorContext& ctx) {
+    auto txIter = WriteInfosToTx.find(sender);
+    Y_ABORT_UNLESS(!txIter.IsEnd());
+
+    auto& tx = (*txIter->second);
+    if (ev) {
+        tx.WriteInfo = std::move(ev);
+    } else {
+        tx.Predicate = false;
+        tx.WriteInfoApplied = true;
+    }
+
+    WriteInfosToTx.erase(txIter);
+    ProcessTxsAndUserActs(ctx);
+}
+
+TPartition::EProcessResult TPartition::ApplyWriteInfoResponse(TTransaction& tx) {
+    Y_ABORT_UNLESS(tx.WriteInfo);
+    auto& srcIdInfo = tx.WriteInfo->SrcIdInfo;
+
+    EProcessResult ret = EProcessResult::ContinueWaitCommit;
+    const auto& knownSourceIds = SourceIdStorage.GetInMemorySourceIds();
+    for (auto& s : srcIdInfo) {
+        auto ins = TxAffectedSourcesIds.insert(s.first).second;
```

Comment:
In case of immediateTx you need to use WriteAffectedSourceIds. The same bug with TxAffectedClients.